### PR TITLE
Enhancements to assembly resolution

### LINKED ISSUE #81: Include more assemblies in host by default
I shouldn't need to do a private binary drop of Microsoft.Azure.WebJobs.dll and Newtonsoft.Json.dll to be able to use these from a function. Can we include these two by default?

### PR DESCRIPTION
**Pull request by fabiocav**
This PR provides the following enhancements:
##### New default assembly references
Added Microsoft.Azure.WebJobs to the list of default assembly references. This allows users to consume types like ICollector<T> without having to add an explicit assembly reference or copy private assemblies.

##### Enhancements to private assembly resolution
Users can now reference private assemblies without a path so what was previously this:
`#r "bin\MyAssembly.dll"` 
Now becomes:
`#r "MyAssembly.dll"` or `#r "MyAssembly.exe"`
We'll continue to use the bin path, but that is where all private assembly probing will take place.

##### Shared assemblies
Introducing a mechanism to allow the host to share its assemblies with the function. This creates an opt-in model to allow functions to request assemblies the host may share.
Shared assemblies are referenced just like framework assemblies, by using their name:
`#r "Newtonsoft.Json"`

> At the moment, Json.NET is the only shared assembly exposed by the host.

##### Additional Imports
Added `Microsoft.Azure.WebJobs` and `Microsoft.Azure.WebJobs.Host` as default namespace imports.
This eliminates the need to add a using clause to reference those namespaces in the functions, making them more concise. This will impact many functions, since whenever logging is required, they need to define an argument of type `TraceWriter`, which required a using statement for the host. That is no longer the case.

Resolves #81

### CODE REVIEW

File: src/WebJobs.Script/Description/CSharp/FunctionMetadataResolver.cs [hunk 1]
```diff
@@ -37,9 +37,15 @@ public sealed class FunctionMetadataResolver : MetadataReferenceResolver
...
+                typeof(IAsyncCollector<>).Assembly.Location, /*Microsoft.Azure.WebJobs*/
+                typeof(JobHost).Assembly.Location, /*Microsoft.Azure.WebJobs.Host*/
+                typeof(CoreJobHostConfigurationExtensions).Assembly.Location, /*Microsoft.Azure.WebJobs.Extensions*/
+                typeof(System.Web.Http.ApiController).Assembly.Location /*System.Web.Http*/
+            };
+
+        private static readonly Assembly[] SharedAssemblies =
+            {
+                typeof(Newtonsoft.Json.JsonConvert).Assembly /*Newtonsoft.Json*/
             };
```
Comment (davidebbo) [inline src/WebJobs.Script/Description/CSharp/FunctionMetadataResolver.cs]: What's the semantic difference between default and shared assemblies?

Comment (davidebbo) [inline src/WebJobs.Script/Description/CSharp/FunctionMetadataResolver.cs]: Oh, I see you explain that in the PR! Maybe add one line comment on SharedAssemblies, e.g. "Assemblies that are not referenced by default but can be explicitly referenced using `#r`"

**Comment by davidebbo** [comment]
:shipit:

File: src/WebJobs.Script/Description/CSharp/FunctionMetadataResolver.cs (hunk 1 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/CSharp/FunctionMetadataResolver.cs]: Default assemblies are automatically added, without the need to explicitly have a "#r" reference. Shared assemblies are the ones we make available to you _if_ you ask for it (e.g. #r "Newtonsoft.Json") but are not automatically injected. This minimizes the number of compile time references that may be unnecessary and make it simpler for you to reference a different version of a given assembly if you want to go that route.

**Comment by mathewc** [comment]
We'll have to be careful not to cause confusion for users. I understand they want to pull in ICollector and other core binding types, but they'll have to know that all the binding attributes don't make sense, since bindings are configured in function.json

**Pull request merged**